src/aspire/basis/fpswf_2d.py

```diff
@@ -1,5 +1,300 @@
+            if num_angular_points[i] % 2 == 1:
+                num_angular_points[i] += 1
+
+        temp = 2 * np.pi / num_angular_points
+
+        t = 2
+
+        quad_rule_radial_weights = temp * radial_quad_points * radial_quad_weights
+        quad_rule_weights = np.repeat(quad_rule_radial_weights, repeats=num_angular_points)
+        quad_rule_pts_r = np.repeat(radial_quad_points, repeats=(num_angular_points / t).astype('int'))
+        quad_rule_pts_theta = np.concatenate([temp[i] * np.arange(num_angular_points[i] / t)
+                                              for i in range(len(radial_quad_points))])
+
+        pts_x = quad_rule_pts_r * np.cos(quad_rule_pts_theta)
+        pts_y = quad_rule_pts_r * np.sin(quad_rule_pts_theta)
+
+        return pts_x, pts_y, quad_rule_weights, radial_quad_points, quad_rule_radial_weights, num_angular_points
+
+    def _generate_pswf_radial_quad(self, n, bandlimit, phi_approximate_error, lambda_max):
+        x, w = leggauss_0_1(20 * n)
+
+        big_n = 0
+
+        x_as_mat = x.reshape((len(x), 1))
+
+        alpha_n, d_vec, approx_length = self.pswf_func2d(big_n, n, bandlimit, phi_approximate_error, x, w)
+
+        cut_indices = np.where(bandlimit / 2 / np.pi * np.absolute(alpha_n) < lambda_max)[0]
+
+        if len(cut_indices) == 0:
+            k = len(alpha_n)
+        else:
+            k = cut_indices[0]
+
+        if k % 2 == 0:
+            k = k + 1
+
+        range_array = np.arange(approx_length).reshape((1, approx_length))
+
+        idx_for_quad_nodes = int((k + 1) / 2)
+        num_quad_pts = idx_for_quad_nodes - 1
+
+        phi_zeros = self.find_initial_nodes(x, n, bandlimit / 2, phi_approximate_error, idx_for_quad_nodes)
+
+        def phi_for_quad_weights(t):
+            return np.dot(t_x_mat2(t, big_n, range_array, approx_length), d_vec[:, :k - 1])
+
+        b = np.dot(w * np.sqrt(x), phi_for_quad_weights(x_as_mat))
+
+        a = phi_for_quad_weights(phi_zeros.reshape((len(phi_zeros), 1))).transpose() * np.sqrt(phi_zeros)
+        init_quad_weights = lstsq(a, b, rcond=None)
+        init_quad_weights = init_quad_weights[0]
+        tolerance = np.spacing(1)
+
+        def obj_func(quad_rule):
+            q = quad_rule.reshape((len(quad_rule), 1))
+            temp = np.dot((phi_for_quad_weights(q[:num_quad_pts]) * np.sqrt(q[:num_quad_pts])).transpose(),
+                      q[num_quad_pts:])
+            temp = temp.reshape(temp.shape[0])
+            return temp - b
+
+        arr_to_send = np.concatenate((phi_zeros, init_quad_weights))
+        quad_rule_final = least_squares(obj_func, arr_to_send, xtol=tolerance, ftol=tolerance, max_nfev=1000)
+        quad_rule_final = quad_rule_final.x
+        quad_rule_pts = quad_rule_final[:num_quad_pts]
+        quad_rule_weights = quad_rule_final[num_quad_pts:]
+        return quad_rule_pts, quad_rule_weights
+
+    def find_initial_nodes(self, x, n, bandlimit, phi_approximate_error, idx_for_quad_nodes):
```

Comment:
Is this a private function? If so, prefix with `_`.

Reply:
Done.